### ISSUE TITLE
httpDelete did not send body payload

### ISSUE THREAD

**Issue by babedev** [kittinunf/Fuel#306]
Fuel: `v1.12.0`

I tried to sent HTTP DELETE with JSON body. I checked `HttpClient.kt` that it skip body set up for GET, DELETE, HEAD

```kotlin
private fun setDoOutput(connection: HttpURLConnection, method: Method) = when (method) {
   Method.GET, Method.DELETE, Method.HEAD -> connection.doOutput = false
   Method.POST, Method.PUT, Method.PATCH -> connection.doOutput = true
}
```

Should we allow `Method.DELETE`? Because I have scenario that API need body payload for DELETE.

Same issue #245

**Comment by kittinunf** [comment]
Yeah, I think by the time I have designed this base on the [RFC](https://tools.ietf.org/html/rfc2616) spec. It is not specified as permitted or not so that I didn't follow through with the implementation for this. However, most of the modern HTTP libraries support this so we should support it too! 

@babedev Would you mind creating a PR for this? I can help you out as well.

**Comment by babedev** [comment]
Sure. I will let it support both GET and DELETE.

**Comment by simonharrer** [comment]
Is there already a PR open for this? I need to interface an API that *requires* me to send a json body alongside a GET request.... :-(

**Comment by kittinunf** [comment]
I haven't get time to work on this yet, sorry. 😢

**Comment by kmcgill88** [comment]
Any movement on this? I need my `DELETE` to have a payload.

**Comment by markGilchrist** [comment]
@kittinunf  do you want me to look at this?

**Comment by kmcgill88** [comment]
Well I shaped up and changed my implementation to not send a body in a `DELETE` per the RFC.

**Comment by markGilchrist** [comment]
@kittinunf @babedev  is there still a need for this as a feature?

**Comment by SleeplessByte** [comment]
Before, the spec defined that the body **SHOULD** be ignored (and therefore a lot of clients did not allow for a `DELETE` request's body, but this was updated with RFC7231:

https://tools.ietf.org/html/rfc7231#section-4.3.5
```
4.3.5.  DELETE
  
   [...]
   
   A payload within a DELETE request message has no defined semantics;
   sending a payload body on a DELETE request might cause some existing
   implementations to reject the request.

   Responses to the DELETE method are not cacheable.  If a DELETE
   request passes through a cache that has one or more stored responses
   for the effective request URI, those stored responses will be
   invalidated (see Section 4.4 of [RFC7234]).
```

Therefore, even though current users might not need this as a feature, it **SHOULD** be implemented.

**Comment by kittinunf** [comment]
Yes, I agree that we should update this to accommodate users' needs.

**Comment by markGilchrist** [comment]
@babedev @simonharrer @kittinunf @SleeplessByte @kmcgill88  Guy I am going to pick this up and open a WIP Pr, I would really appreciate you feed back and reviews.

**Comment by eybeestudios** [comment]
is there an estimated time when this will be merged? i just recognized that i use an API that needs a request-body for a DELETE and i dont want to change the lib for this single request.

**Comment by iNoles** [comment]
@markGilchrist In my opinion, doOutput should be configurable to any developers.

**Comment by markGilchrist** [comment]
@eybeestudios  this is now merged in

**Comment by simonharrer** [comment]
@markGilchrist Thanks for your effort! Had a look at your PR - nice work btw., but are you sure that GET requests can have a payload now as well? From the PR #417 I can only see the change for the DELETE requests.

**Comment by markGilchrist** [comment]
@simonharrer No that work was not included in this PR it only covers DELETE, I am working on a second PR for GET, if I tag it in here I would really appreciate any comments you may have!

**Comment by SleeplessByte** [comment]
Seems that this can be closed now, and we can continue on the PR once it's up ;)

**Comment by smallufo** [comment]
Hi , is this issue really solved ? I try to fire an DELETE request (with body) to FB's API , but it seems it dosen't work.

sample code : 



    val uri = "https://graph.facebook.com/$version/me/messenger_profile?access_token=$pageToken"

    val root = json {
      obj("fields" to array(PERSISTENT_MENU))
    }
    val json = klaxon.toJsonString(root)
    logger.info("delete json = {}" , json)

    val (req: Request, res: Response, result: Result<String, FuelError>) = Fuel.delete(uri)
      .header(mapOf("Content-Type" to "application/json"))
      .body(json)
      .responseString()

    logger.info("req string = {}" , req.toString())


But FB seems not receiving the body . See the attachment 
![2018-10-28_21-53-37](https://user-images.githubusercontent.com/626659/47616791-653e5e00-dafc-11e8-84f3-c58fba3c58ed.png)


But the same code works in curl : 

![2018-10-28_22-00-23](https://user-images.githubusercontent.com/626659/47616852-00373800-dafd-11e8-85b2-39c91fc54026.png)

I am confused , what may go wrong here ?

**Comment by SleeplessByte** [comment]
```kotlin
val version = 3
val pageToken = UUID.randomUUID().toString()
val uri = "https://graph.facebook.com/$version/me/messenger_profile?access_token=$pageToken"

val root = "{ \"fields\": [\"$PERSISTENT_MENU\"] }"

val request = Fuel.delete(uri)
  .header(mapOf(Headers.CONTENT_TYPE to "application/json"))
  .body(root)

println(request.cUrlString())
```

This gives me 
```bash
$ curl -i -X DELETE -d "{ \"fields\": [\"persistent_menu\"] }" -H "Content-Type:application/json" https://graph.facebook.com/3/me/messenger_profile?access_token=7940c9cc-656c-42c3-9d9d-f786bd30699d
```

I've also tested this via a mocked request, and it seems fine :/

What version are you on?

**Comment by smallufo** [comment]
I am using Fuel 1.15.0 , with FB's graph version `v3.2`
BTW , I don't think your code works , because it's not a valid page_access_token.
If it passes access token check , it will check the content, which leads it cannot find my `fields` field

Is there anyway to truly dump the network layer stream with Fuel's API ? I am not sure if it FB or Fuel's implementation problem...

**Comment by SleeplessByte** [comment]
Obviously I wasn't going to post my actual token hehe ;)

That said, everything is piped to HttpURLConnection and then it leaves fuel. No real way of intercepting it at that link without using a proxy.

Can you print your request curlstring and see it that one works?

**Comment by smallufo** [comment]
Wow , I don't know Fuel has `curlString` function , great. This is my result : 

`curl -i -X DELETE -d "{\"fields\": [\"persistent_menu\"]}" -H "Accept-Encoding:compress;q=0.5, gzip;q=1.0" -H "Content-Type:application/json" https://graph.facebook.com/v3.2/me/messenger_profile?access_token=EAAGpXXX`


![intellij](https://user-images.githubusercontent.com/626659/47619055-35e81b00-db15-11e8-8191-d5e42be34c2e.png)
The result is the same.

BUT , very interesting . After I copy the curlString to console and execute , it is OK !!!
![curl](https://user-images.githubusercontent.com/626659/47619066-51ebbc80-db15-11e8-9ca2-53a084a98176.png)

This is so weird.

**Comment by SleeplessByte** [comment]
That's actually really helpful.

I can think of a few things, but for now, can you also check this slightly
weird thing:

```kotlin
val (request, _, _) = what you already have.response()
println(request.cUrlString())
```

My initial thought is that Facebook is actually choking on something else,
or that this still has the content type double header bug.

**Comment by smallufo** [comment]
I try to remove the `Accept-Encoding` header , 

    val (req: Request, res: Response, result: Result<String, FuelError>) = Fuel.delete(uri).also {
      it.headers.remove("Accept-Encoding")
    }
      .header(mapOf("Content-Type" to "application/json"))
      .body(json)
      .responseString()

    logger.info("curlString = {}" , req.cUrlString())


which generates this : 

`curl -i -X DELETE -d "{\"fields\": [\"persistent_menu\"]}" -H "Content-Type:application/json" https://graph.facebook.com/v3.2/me/messenger_profile?access_token=EAAGpxxx
`

The result is the same : failed in java code , success in curl console.

If I remove `Content-Type` header (which leads empty header) , it won't pass FB (both code and cURL) . So  `Content-Type` header is mandatory here.

**Comment by SleeplessByte** [comment]
Alright. This might be an issue on our side. I will try some things out over here and see if I can make this work (and write a test so it doesn't fail again/regression)

**Comment by SleeplessByte** [comment]
It doesn't.

**Comment by checketts** [comment]
Just as a datapoint, RequestBody with DELETE wasn't working for me on `1.13.0`, and after upgrading to `1.16.0` it is working for me.  I'm not discounting @smallufo 's report, just letting you know that in my case it is working.

**Comment by SleeplessByte** [comment]
Thank you @checketts, that does help. I still think it's something with Facebook, but I now have my own page set up so I can debug this for @smallufo .

**Comment by kittinunf** [comment]
@smallufo is there any update on this? Do you think it is still an issue?
